### PR TITLE
[WIP] Feature/patch/softmax cross entropy with logits

### PR DESCRIPTION
**Pull request by MFreidank**
Addressing previously reported [tensorflow issue #38185](https://github.com/tensorflow/tensorflow/issues/38185). Also tied to `tfdeterminism` issues #19, #14, #9. 
Fused implementation of `{sparse_,}softmax_cross_entropy_with_logits` exhibits non-determinism in its backprop. 
This PR will provide a patch that maintains the same function interface but computes the computational steps involved in a non-fused way that is fully deterministic.

### CODE REVIEW

File: tfdeterminism/patch.py [hunk 1]
```diff
@@ -70,13 +70,20 @@ def _patch():
   if re.match("(1\.(14|15)|2\.0)", tf_version):
     os.environ['TF_CUDNN_DETERMINISTIC'] = '1'
     _patch_bias_add()
+    _patch_fused_softmax_cross_entropy()
```
Comment (duncanriach) [inline tfdeterminism/patch.py:73]: I'll push changes that rough-out `enable_determinism` and then you can call this from the appropriate part of that.

File: tfdeterminism/patch.py [hunk 2]
```diff
@@ -135,3 +142,110 @@ def _new_bias_add_1_14(value, bias, data_format=None, name=None):
...
+      energies are interpreted as unnormalized log probabilities.
+    dim: The class dimension. Defaulted to -1 which is the last dimension.
+    name: A name for the operation (optional).
+    axis: Alias for dim.
+  Returns:
+    A `Tensor` that contains the softmax cross entropy loss. Its type is the
+    same as `logits` and its shape is the same as `labels` except that it does
+    not have the last dimension of `labels`.
+  """
+  raise NotImplementedError()
```
Comment (duncanriach) [inline tfdeterminism/patch.py:200]: I like the baby steps. Good job!

**Comment by duncanriach** [comment]
Comments above. Heads-up, before you're done, please remember to implement `test/test_softmax_cross_entropy.py` in the spirit of `test/test_patch_bias_add.py`. This will use the appropriate forward code from the upstream TensorFlow repo along with the deterministic gradient code that I provided in TensorFlow [Issue 38185](https://github.com/tensorflow/tensorflow/issues/38185) (also testing all the modified op bindings).

File: tfdeterminism/patch.py [hunk 3]
```diff
@@ -135,3 +142,110 @@ def _new_bias_add_1_14(value, bias, data_format=None, name=None):
           value, array_ops.reshape(bias, broadcast_shape), name=name)
     else: # data_format == 'NHWC' or data_format == None
       return math_ops.add(value, bias, name=name)
+
+
+def _patch_fused_softmax_cross_entropy():
+  # Sparse
+  tf.nn.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # access via public API
+  nn.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # called from tf.keras.layers.convolutional.Conv
+  nn_ops.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # called from tests
```
Comment (duncanriach) [inline tfdeterminism/patch.py:151]: Looking at [tensorflow/python/ops/nn_ops.py](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/nn_ops.py#L3222) it seems like it might be necessary to remap the bindings for `softmax_cross_entropy_with_logits_v2` as well. After a brief review, I'm not sure what the difference between this and the other versions is, though, apart from the deprecated `dim` parameter.

File: tfdeterminism/patch.py [hunk 4]
```diff
@@ -135,3 +142,110 @@ def _new_bias_add_1_14(value, bias, data_format=None, name=None):
...
+
+  # Non-sparse
+  tf.nn.sparse_softmax_cross_entropy_with_logits = _new_sparse_softmax_cross_entropy_with_logits_1_14  # access via public API
+  nn.sparse_softmax_cross_entropy_with_logits = _new_sparse_softmax_cross_entropy_with_logits_1_14  # called from tf.keras.layers.convolutional.Conv
+  nn_ops.sparse_softmax_cross_entropy_with_logits = _new_sparse_softmax_cross_entropy_with_logits_1_14  # called from tests
+
+
+# The original, pre-patched method can be viewed at
+# https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/ops/nn_ops.py#L3182
+def _new_softmax_cross_entropy_with_logits_1_14(labels, logits, axis=-1, name=None):
```
Comment (duncanriach) [inline tfdeterminism/patch.py:161]: BTW, the reason that _new_bias_add_1_14 has the version number in the name is that before stock TensorFlow version 1.14, `tf.nn.bias_add` has a different API. I have a version of this patch for version 1.13 and earlier that I didn't release because it was not useful without `TF_CUDNN_DETERMINISTIC`, which was released in version 1.14 of TensorFlow.

I don't think we need the `_1_14` on the end of these method names. However, before enabling this patch to be applied to any given version of TensorFlow (as decided in `enable_determinism`) it should be confirmed that the patched version of the method has the same API as the original on in that version of TensorFlow.

File: tfdeterminism/patch.py [hunk 5]
```diff
@@ -135,3 +142,110 @@ def _new_bias_add_1_14(value, bias, data_format=None, name=None):
...
+
+def _patch_fused_softmax_cross_entropy():
+  # Sparse
+  tf.nn.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # access via public API
+  nn.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # called from tf.keras.layers.convolutional.Conv
+  nn_ops.softmax_cross_entropy_with_logits = _new_softmax_cross_entropy_with_logits_1_14  # called from tests
+
+  # Non-sparse
+  tf.nn.sparse_softmax_cross_entropy_with_logits = _new_sparse_softmax_cross_entropy_with_logits_1_14  # access via public API
+  nn.sparse_softmax_cross_entropy_with_logits = _new_sparse_softmax_cross_entropy_with_logits_1_14  # called from tf.keras.layers.convolutional.Conv
```
Comment (duncanriach) [inline tfdeterminism/patch.py:155]: The comment on the end of this line is not relevant here (same for the non-sparse version). You should search for calls to `sparse_softmax_cross_entropy_with_logits` (and the non-sparse version) in `tensorflow/python` and do the re-binding as appropriate for the op (with similar, perhaps in some cases identical, comments about why you're doing it).

**Comment by duncanriach** [comment]
I just added `enable_determinism` with [this commit](https://github.com/NVIDIA/tensorflow-determinism/commit/0c5aac5bc25eb71099aa6b3ae98ce6b270988c4a). You should use `enable_determinism` to apply the patch in `test/test_patch_fused_softmax_cross_entropy.py`. I'm going reorganize the testing quite a lot before release. For the testing part, I would like you to focus on getting `./container.sh <container-image> python test_patch_fused_softmax_cross_entropy.py` complete and working on `tensorflow/tensorflow:1.14.0-gpu` (TF1 API) and `tensorflow/tensorflow:2.2.0-gpu` (TF2 API).

**Comment by duncanriach** [comment]
Hi @MFreidank, you haven't resolved the conflicts correctly.

It looks like what you've done is merged my changes in master into your branch so that it's a mixture of your changes and my changes. The _files changed_ tab in this current pull request shows mostly my changes as if they were your changes.

Instead, you should `git rebase` your incremental commits in your working branch to the most recent state of the master branch (after syncing your fork's master branch to the latest upstream master) and then `git push --force` them to your fork. The commits in the branch you seek to merge should only represent incremental, non-conflicting, changes on top of the latest state of the master branch. This will result in the _files changed_ report for the pull request showing only your incremental changes over master.

File: tfdeterminism/enable_determinism.py [hunk 6]
```diff
@@ -52,7 +52,7 @@ def _enable_determinism(seed=None):
     _patch_bias_add()
   if in_ngc_cont and ngc_vers.at_least('19.06') or tf_vers.at_least('2.1'):
     os.environ['TF_DETERMINISTIC_OPS'] = '1'
+    # TODO: Add patch crossentropy here as well? Issue seems to still be present on tf 2.1, 2.2
```
Comment (duncanriach) [inline tfdeterminism/enable_determinism.py:55]: No. This is the condition for setting `TF_DETERMINISTIC_OPS=1`: NGC containers with version >= 19.06 or stock TensorFlow with version >= 2.1.

The condition below will ensure that the fused softmax/cross-entropy patch is applied to NGC containers with version >= 19.06 or stock TensorFlow with version >= 1.14 (which includes versions 2.1 and 2.2).

File: tfdeterminism/patch.py [hunk 7]
```diff
@@ -70,14 +70,20 @@ def _patch():
   if re.match("(1\.(14|15)|2\.0)", tf_version):
     os.environ['TF_CUDNN_DETERMINISTIC'] = '1'
     _patch_bias_add()
-    # Apply the fused softmax/cross-entropy patch here
+    _patch_fused_softmax_cross_entropy()
+    print("TensorFlow version %s has been patched "
+          "using tfdeterminism version %s" %
+          (tf_version, __version__), file=sys.stderr)
+  elif re.match("2\.1|2\.2", tf_version):
```
Comment (duncanriach) [inline tfdeterminism/patch.py:77]: I want to deprecate patch, not update it, and no longer advertise it in the documentation. If folks want determinism in TensorFlow version 2.1 or 2.2 then they should be using `enable_determinism`. Not adding this does not break anything because there was no patch available for TensorFlow versions 2.1 and 2.2 before.

**Comment by duncanriach** [comment]
I accidentally committed and pushed only part of the changes I made today, so I did a somewhat risky force push to master to amend the commit. I did this hoping that nobody pulled between the two pushes (which is very unlikely). This force push seems to have enabled/forced GitHub to re-evaluate your merged changes and remove my commit from your pull request. Your pull request makes more sense now, but still has replicated commits and a merge commit.

I made some more conflicting changes. I still recommend that you rebase your branch to upstream/master, remove replicated and merge commits, and force push your branch.

**Comment by phqtuyen** [comment]
Hi, my name is Thomas, I am working at the Oracle Digital Assistant team in which we made extensively use of Tensorflow and Tensorflow-determinism, we really appreciate this wonderful library. I just wish to ask is this issue going to be merged anytime soon?

**Comment by duncanriach** [comment]
Hi Thomas (@phqtuyen),

Thanks for letting us know. These ops seem to be generally high priority for determinism in TensorFlow.

@MFreidank: please can you give an update on this and/or move it forward. Most of the rest of the infrastructure is now in place for this.

**Comment by MFreidank** [comment]
Hi @phqtuyen, @duncanriach 
Sorry for the wait, I had been pulled into other things. 
I've picked up work on this now and will have an update to the PR before end of this week. 
I'll base my changes on the new infrastructure recently put in place.

**Comment by duncanriach** [comment]
Awesome. Just FYI, I'll be offline next week (2020-08-17 through 2020-08-21) and, since I'll need to be involved in reviews and iterations, that might slow things down a bit.

**Comment by duncanriach** [comment]
Hey @MFreidank, are you still planning on on completing this PR?

**Comment by duncanriach** [comment]
Hi @Freidank, I'm about to go offline again until 2020-09-16, but I wanted to check-in with you again. I think that this op's nondeterminism is near the top of the priority/urgency list for TensorFlow, and I want to get it resolved ASAP. Can you commit to pushing this through from your end?

**Comment by MFreidank** [comment]
Hi Duncan,

Sorry for the delay on this.
Yes, I'll pick it up starting tomorrow my time.

Duncan Riach <notifications@github.com> schrieb am Fr., 4. Sep. 2020, 20:07:

**Comment by duncanriach** [comment]
Thank you, Moritz.

**Comment by MFreidank** [comment]
Hi again @duncanriach,
I've implemented a first version (on top of the most recent `master` branch with `enable_determinism`) that patches only `tf.nn.sparse_softmax_cross_entropy_with_logits` and `tf.nn.softmax_cross_entropy_with_logits` and integrated the unit test case you shared into `test/test_softmax_cross_entropy_with_logits.py`. 
To test that this works as intended I'll need access to my GPU which will only be available to me in another 6 hours. 
If the test cases pass, I'll update the PR with the above.
What remains to be determined is which other functions need to be patched after that (by scanning through `tensorflow/python` as outlined by you above). 
I will first focus on getting the minimal patch for the functions in `tf.nn` to work including unit tests and will then broaden it to include the other functions.

**Comment by duncanriach** [comment]
Thanks, @MFreidank. Sounds good. Just to reiterate, I'll be offline until 2020-09-16 and I'll review this after that.

**Comment by duncanriach** [comment]
Hey @MFreidank, please will you submit those changes so that I can review them?

**Comment by duncanriach** [comment]
@MFreidank, we will soon release version 0.4.0. We intend to include the fused softmax/cross-entropy patch in version 0.5.0 and I want to get started on it ASAP. If I don't hear from you in the 24 hours, we will proceed to implement it without you.
